### DISCUSSION TITLE
Shouldn't an out-of-date dependency trigger an otherwise up-to-date task?

### DISCUSSION THREAD

**Discussion by ConX** [go-task/task#669] [Q&A] [answered]
Hi everyone,

First of all, thanks for the great tool!

I just noticed that when a dependency is out-of-date, that doesn't necessarily trigger the target task. The scenario that I have is that of a target task (T) that generates some files and requires a file generated by another task (A). The way I tried to implement that was by listing the task (A) as a dependency of (T) and also listing sources and generates for the task (T). I expected that if task (A) is out-of-date, that will trigger task (T) as well, but that isn't the case.

I assume that I can list the file generated by task (A) in the sources of (T), and that will fix the issue, but I am curious why this isn't implied, given that (A) is a dependency of (T). What is a scenario for which the current approach is desired?

Here is a minimal reproducible example that uses `status`  instead of files:

```
version: "3"

tasks:
  first-dep:
    cmds:
      - echo First
    status:
      - "exit 1"

  second-dep:
    cmds:
      - echo Second
    status:
      - "exit 0"

  target:
    cmds:
      - "echo Target"
    deps:
      - first-dep
      - second-dep
    status:
      - "exit 0"
```

and here is the output that I get:

```
$task target
task: Task "second-dep" is up to date
task: [first-dep] echo First
First
task: Task "target" is up to date
```

**Comment by ConX** [comment]
I forgot to mention that I would work on a PR if allowing for such behavior (other by default or through an option) is something others find useful.

**Reply by ghostsquad** [reply]
I believe that whether or not a dependency changes is irrelevant unless you specify that it is relevant. The default behavior is to run a task everytime unless you tell go-task that it shouldn't run via status. The example provided is unfortunately not very useful to me to really dig deeper into this use case.

Here's an example that may be useful:

TaskA generates fileA
TaskB consumes fileA and generates fileB

In this case, taskB does in fact depend on the output of taskA via fileA.

TaskB should then use FileA as a way to determine if it should run.

Do you have another scenario that we can explore together?

**Reply by ghostsquad** [reply]
Maybe the issue is that 99% of the time, if a dependency changes, the current task needs to run? I'm not sure that's true.

If it is true, then setting up status/sources/generates becomes tedious boilerplate.

I am very curious about the specific situation you are trying to optimize.

**Reply by ConX** [reply]
I agree that when all tasks generate a single output file and have a list of file sources it's usually not unreasonable to write a chain of `sources` and `generates` statements to achieve the dependencies. Though, having to do that manually, especially in long chains of tasks, I believe diminishes the power of the `deps` which can serve as an abstraction mechanism, that does not require tracking file names (to populate `sources` and `generates`), but rather using the user-friendly custom task aliases.

Where it starts becoming tedious is when tasks use programmatic checks to determine if they should run or not and do not generate files.  

Here is a three-task pseudo-code example for demonstration:
```
task archive:
  cmd: tar to archive files
  sources: src_file1, src_file2
  generates target.tar.gz
task build-cloud-instance-image:
  cmd: <cloud image builder>
  sources target.tar.gz
task cloud-deploy:
  cmd: <deploy cloud instance with latest image>
  status: check if an instance already exists
  deps: build-cloud-instance-image
```

In this case above, the first time when `task cloud-deploy` is invoked everything will run as expected. Now, let's assume that we update the file `src_file1` and invoke `task cloud-deploy` again. This would make tasks `archive` and `build-cloud-instance-image` run but since the instance already exists task `cloud-deploy` will be skipped. With the current implementation, this can be fixed by adding the `target.tar.gz` file in the sources of `cloud-deploy` or by programmatically tracking if the image has been updated since the last run. My original expectation of `deps` was that since one or more dependencies had to be executed so should the current task.

> Maybe the issue is that 99% of the time, if a dependency changes, the current task needs to run? I'm not sure that's true.

I am also not sure! I just can't think of a scenario where it would be desired for the current task to be skipped when a dependency has changed. Moreover, there is a already mechanism to support this behavior that better reflects the use of the terms "current task" and "dependency." For example, in the case above, if I didn't want the `cloud-deploy` to run, and I only wanted to rebuild the cloud instance image, I would simply invoke `task build-cloud-instance-image` instead.

**Reply by ghostsquad** [reply]
Right, so I think you got the nail on the head, but maybe didn't realize it. `cloud-deploy` should be an idempotent operation.

Meaning, I would explicitly _not_ give this a `status`. The script that deploys the instance should have all the logic in place to do the following:

1. If the instance does not exist, create it
2. If the instance already exists, check if the relevant source files have changed.
2a. If the source files or other attributes have changed, perform an operation to bring the deployed instance in line with new desired state.
2b. If desired state (source files) match what has been deployed, exit with code 0 (and a helpful message that everything is up-to-date)

If you are familiar with Terraform or Pulumi, this is what they do when you run `apply` or `up`, respectively.

Kubernetes also does this operation continuously. The manifests applied represent the desired state. Kubernetes is a reconciliation engine that is constantly evaluating the current state and trying to converge it to the desired state.

You _could_ use go-task in a similar way, by decoupling the scripts to check if an update is needed, from the scripts that update the instance, but I don't think that's actually the best idea. There will likely be duplication in that, and the result would be the same anyways.

TLDR: make `cloud-deploy` idempotent.

**Reply by ConX** [reply]
That's why I initially tried to avoid giving a specific example, as my goal wasn't to find a solution to my specific case.

You must agree that all of what you describe above can be more easily resolved by having a change in one of the dependencies, trigger the dependent tasks.

The reason I started this discussion topic, is to understand why this isn't the current behavior and what are the downsides of making that change, if any. Could you provide an equivalent real-life example where the current behavior of skipping the target task is desired despite the execution of its dependencies?

**Reply by ConX** [reply]
@ghostsquad Expanding from the example you gave in your first comment, below is another code that I hope better illustrates the point I am trying to make.

```
version: "3"

tasks:
  create-dummy-files:
    cmds:
      - |
        for i in `seq 5`; do
          echo "${i}" >> "file${i}"
        done;

  modify-file1:
    cmds:
      - echo {{.RANDOM}} > file1
    vars:
      RANDOM:
        sh: date +%N|sed s/...$//

  task1:
    cmds:
      - cat file1 file2 > file6
    sources:
      - file1
      - file2
    generates:
      - file6

  task2:
    cmds:
      - cat file3 file4 > file7
    sources:
      - file3
      - file4
    generates:
      - file7

  task3:
    cmds:
      - cat file5 file6 file7 > file8
    sources:
      - file5
    deps:
      - task1
      - task2
```

The first two tasks are there just for convenience generating the files. The demo would be something like this:

1. `task create-dummy-files`
2. `task task3` (Should run everything as expected)
3. `task task3` (Everything gets skipped as it's up-to-date)
4. `task modify-file1`
5. `task task3` (Should run `task1` and `task3`)

My question is why does `task3` require both the `deps` and the `sources` to be defined to achieve this?  I understand that `deps` is necessary in order for `task1` and `task2` to be evaluated and run if needed. But doesn't that alone infer _"in order to run `task3` you need whatever `task1` and `task2` are generating/doing?"_ If so, why the redundancy of having to specify the sources? If not, what is an example of where the dependencies would need to be executed, but the target task should be skipped?

**Reply by ghostsquad** [reply]
Ya, you are right. If you look at traditional Makefile, dependent targets, if changed will cause parent targets to trigger.

Status and Sources is just a way to give flexibility as to how to define if _this_ target should run, assuming it has no dependencies.

> The following Makefile has three separate rules. When you run make blah in the terminal, it will build a program called blah in a series of steps:
>
> Make is given blah as the target, so it first searches for this target
>
> blah requires blah.o, so make searches for the blah.o target
>
> blah.o requires blah.c, so make searches for the blah.c target
>
> blah.c has no dependencies, so the echo command is run
>
> The cc -c command is then run, because all of the blah.o dependencies are finished
>
> The top cc command is run, because all the blah dependencies are finished
>
> That's it: blah is a compiled c program


```makefile
blah: blah.o 	cc blah.o -o blah # Runs third


blah.o: blah.c 	cc -c blah.c -o blah.o # Runs second

blah.c: 	echo "int main() { return 0; }" > blah.c # Runs first
```

https://makefiletutorial.com/

I think this is a legitimate bug now. Thanks for helping me understand this.

**Reply by ghostsquad** [reply]
The question now:

How would you indicate that the parent target (`task3` in your example) _*should*_ run even if it's deps are up-to-date?

How would you indicate that the parent target (`task3` in your example) _*should not*_ run if it's deps are up-to-date?

I'd assume also that tasks under cmd should be considered too? How does that work exactly?

Does status/sources apply to all the elements in the `cmds` array?

If deps or cmds are up-to-date, and you  want to force a run everytime (the current behavior if you exclude sources/status), I think you would need to add:

```
method: none
```

I'd have to double check the code though.

Although I definitely think this likely should be the default behavior, it is not backwards compatible. 🙁

It would likely need to be part of a version 4 if this were to be the new default behavior.

**Reply by ConX** [reply]
Thanks for spending the time! I will attempt to answer the questions although the devs know best:

> How would you indicate that the parent target (`task3` in your example) __should__ run even if it's deps are up-to-date?

That would be based on the `status/sources/generates` statements of `task3`. Although, as stated, I think that including `file5` and `file6`  from the previous example in the `sources` is redundant, there could be other files (not generated by the dependencies) that when included in the `sources` should determine whether `task3` should run. The same goes for `generates` and `status` statements. If none of the `status/sources/generates` is provided irrespective of whether the dependencies are up-to-date or not, `task3` should always run when invoked. 

> How would you indicate that the parent target (`task3` in your example) __should not__ run if it's deps are up-to-date?

Following from above, if dependencies are up-to-date and all the provided `sources/generates/status` resolve to determine `task3` an up-to-date task then that should also be skipped.

> I'd assume also that tasks under cmd should be considered too? How does that work exactly?
> Does status/sources apply to all the elements in the `cmds` array?

That's a good question I hadn't thought of! I view task under cmd more as shortcuts/functions rather than dependencies, thus I would expect those to **not** be considered, but I might be off here.

> If deps or cmds are up-to-date, and you want to force a run everytime (the current behavior if you exclude sources/status), I think you would need to add:

As I wrote above, I think the current approach of excluding the `sources/status/generates` should still force the task to run everytime. Am I missing something?

A pseudo-code of how the resolution of the target/parent task (`task3` in the example) I think should be handled:

1. Evaluate the dependencies of the target task. If any of the dependencies was out-of-date mark the target task as out-of-date (or tainted per Terraform).
2. When returning to evaluate `task3`, if it hasn't already been marked as out-of-date, evaluate its `sources/generates/status` as normal to determine if it should be skipped or not.

To keep the above pseudo-code simple I avoided talking about the dependencies of the dependencies, etc. These should be probably be handled by recursion where each dependency only affects its parents/ancestors. This is probably easier to explain graphically:

![image](https://user-images.githubusercontent.com/338764/155872659-320d8001-8798-4c0a-8c97-43c443fa9f96.png)

For example, if only tasks `dep1.2.1` and `dep2.2` are out-of-date, this would also force tasks `dep1.2`, `dep1`, `dep2`, and of course the `target` to be marked as out-of-date and run.

**Reply by ghostsquad** [reply]
Regarding CMD, the only way to run target deps sequentially would be with cmd.

I feel like deps should be a multi dimensional array. Outer array runs parallel, inner array(s) run sequentially, and Cmds should no longer support the `task:` syntax. I think that would improve readability too.

Anyways back to deps triggering a run...
Even if this change isn't backwards compatible, it feels like a bug. I would support this change.

**Reply by andreynering** [reply]
> I feel like deps should be a multi dimensional array. Outer array runs parallel, inner array(s) run sequentially, and Cmds should no longer support the task: syntax

There are two reasons to keep this as it works today:

1) Changing that would be backward incompatible
2) `task:` is very useful because you can run them in whatever order you want, even mixed with commands. Dependencies always run first. Also, not everyone use `task:` for dependencies, sometimes they are used as helper tasks to run many tasks in sequence, etc

**Comment by andreynering** [comment, marked as answer]
Hi @ConX, thanks for reaching out (and @ghostsquad for the good inputs).

Regarding this question:

> Could you provide an equivalent real-life example where the current behavior of skipping the target task is desired despite the execution of its dependencies?

I think [this web app example](https://github.com/go-task/examples/blob/master/go-web-app/Taskfile.yml), although a very basic example, explains what I had in mind when I initially implemented this feature:

In this case `js` and `css` are dependencies of `assets`, which is a dependency of `build`. If you look at the `sources` of each task, you'll see it'll automatically work as you expect, because each task has as a source the generated files of its dependencies.

It never come to my mind that someone would want to trigger a run when a dependency change in a scenario where you *wouldn't* add the generated files of the dependencies as the task source.

Seems that I was wrong, because with DevOps/Infra stuff you not always deal with files. Also, in the beginning `status` wasn't a thing, and only `sources` and `generates` existed. I also didn't know Makefiles behaved this they do.

---

All that said my thoughts are:

- There may be challenges on trying to implement this, given the way the code works currently
- Since this could possibly be backward incompatible, we probably need a global option in the Taskfile for this for now. We could consider enabling it by default on v4, and perhaps keep the option to disable this behavior
- I think it's worth opening a proper issue for this feature request (given this is a discussion). This way we could track 👍s to know if more people are interested in this before implementing it

**Reply by ghostsquad** [reply]
I've been browsing the existing issues, code base, and also using some of the my own experience from learning the tool, and I am interested in a v4. I think there's a few things that could be done to really polish the tool. I will be trying to put these thoughts and get some more ideas down in a fork in the coming weeks.

For now, I think it's best to make tasks idempotent. If you run into a limitation like this, move the logic of deciding when or when not to act, away from go-task.

**Reply by ConX** [reply]
Hi @andreynering, thanks for your detailed response. I can create an issue based on this discussion to track the popularity of this feature.

**Reply by pavetok** [reply]
It could be a local option for dependency declaration. For example

```
deps:
  - task: foo
    outdated: true
```

**Reply by solvingj** [reply]
I like @pavetok suggestion for the UX.  I originally came looking for solution with the idea that the option should be something like:
-  add the "generates" paths of my dependencies to my "sources"
- OR:  add the "sources" of my dependencies to my "sourcess"
- @pavetok suggestion is much simpler and more broad: "If my dependency is outdated, i am also outdated"

Maybe each of these approaches has a valid use case?

In any case, if nobody is working on this, please do.

**Comment by MoritzWeber0** [comment]
Hi together! I have a similar use case. I have several Dockerimages that depend on each other. 
Task A is a `docker build` command that produces the image `test`. Task B uses the Dockerimage `test` from Task A as base image. Therefore, Task B has to be rebuild if the Dockerimage of Task A changes. 

I have a small workaround for it. I added the checksum file (`.task/checksum/task-a`) as source to Task B. Task B does then always rerun if the sources of Task A changed.

**Comment by solvingj** [comment]
@ConX did you ever open an issue for this?
@andreynering did you ever remember seeing this feature request come through, or implement the idea otherwise?

**Comment by solvingj** [comment]
@ConX @MoritzWeber0 @andreynering :
I submitted https://github.com/go-task/task/pull/2676
if either of you are still interested in this, please let know if this helps with your cases. 
@pavetok would be interested in your thoughts on this as an intermediate step. 
If this goes well, I would make an attempt at adding a similar "include_deps:" flag on the `dep` data structure.

**Comment by solvingj** [comment]
I've added this one instead, based on the original discussion:
https://github.com/go-task/task/pull/2677
